### PR TITLE
Add a cloud renderer that uploads geometry to the GPU, rendering much faster than vanilla.

### PR DESCRIPTION
**Pull request by Zaggy1024**
In the profiler, this new renderer gets ~0.47% of gameRenderer time when I have a 16 chunk render distance, as opposed to vanilla clouds' ~9-11%.

It may be preferable to make this use WorldProvider.cloudRenderer instead, but I wasn't sure where to register the renderer class from to handle WorldEvent.Load (assuming this is the best way to set cloudRenderer).

### CODE REVIEW

**Comment by Zaggy1024** [comment]
Updated to use an enum singleton and have a config entry to disable the feature.

Side note, this renderer should work on all hardware that vanilla Minecraft works on, so if you test it and it doesn't work, let me know.

File: src/main/java/net/minecraftforge/client/FancyCloudRenderer.java [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+            buffer.limit(FORMAT.getNextOffset());
+            for (int i = 0; i < FORMAT.getElementCount(); i++)
+                FORMAT.getElements().get(i).getUsage().preDraw(FORMAT, i, FORMAT.getNextOffset(), buffer);
+            buffer.position(0);
+        }
+
+        /* Passes:
+        0 - Write depth buffer to prevent insides rendering,
+        1 - Render clouds */
+        for (int pass = 0; pass < (WIREFRAME ? 3 : 2); pass++)
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/client/FancyCloudRenderer.java]: Why in the world is this a for-loop? Just write the `case` statements one after the other with an if around the 3rd one. This is just... confusing.

**Comment by asiekierka** [comment]
I feel like this should be a separate mod, while Forge should only have the necessary hooks.

**Comment by Zaggy1024** [comment]
I'll leave the decision on whether to include it in Forge up to Lex and fry, but I feel that having some of the most extreme render time hogs fixed would be a big benefit. Some people on IRC have noted they always turn the cloud setting to fast because the fancy clouds are too slow, but this makes fast and fancy clouds perform almost exactly the same. It may also be worth noting that this implementation is better than Optifine's, since Optifine just updates a display list every 20 ticks with injected hooks around the vanilla renderer.

Indeed, it hadn't occurred to me to target the 1.9.4 branch. Is 1.9.4 in a state where that would be a good thing to do?

**Comment by asiekierka** [comment]
Absolutely.

**Comment by Actuarius** [comment]
@williewillus added labels [Performance]

**Comment by mezz** [comment]
Master is what you should target, it's merged with 1.9.4 now.
This should be updated to resolve the conflicts.

File: src/main/java/net/minecraftforge/common/ForgeModContainer.java [hunk 2]
```diff
@@ -288,9 +295,10 @@ public static void updateNag()
     @SubscribeEvent
     public void onConfigChanged(OnConfigChangedEvent event)
     {
-        if (getMetadata().modId.equals(event.getModID()) && !event.isWorldRunning())
+        if (getMetadata().modId.equals(event.getModID()))// && !event.isWorldRunning())
```
Comment (mezz) [inline src/main/java/net/minecraftforge/common/ForgeModContainer.java]: This doesn't look like it's left here intentionally

File: src/main/java/net/minecraftforge/client/CloudRenderer.java [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+    {
+        return cloudMode == 2 ? 12 : 8;
+    }
+
+    private void vertices(VertexBuffer buffer)
+    {
+        boolean fancy = cloudMode == 2;    // Defines whether to hide all but the bottom.
+
+        float scale = getScale();
+        float CULL_DIST = 2 * scale;
```
Comment (kashike) [inline src/main/java/net/minecraftforge/client/CloudRenderer.java]: `cullDist`? (why upper and underscore?)

**Comment by Zaggy1024** [comment]
Fixed the issues noted by @mezz and @kashike, and the branch is now up to date and ready to merge.

**Comment by williewillus** [comment]
Does this actually replace vanilla clouds if enabled? Where does it do that, I must've missed it?

**Comment by Zaggy1024** [comment]
Whoops, forgot to run genPatches. :|

**Comment by Zaggy1024** [comment]
Done. :)

**Comment by williewillus** [comment]
Rebase to 1.10.x @Zaggy1024? Would love to see this in

**Comment by Zaggy1024** [comment]
I'd rather have confirmation that this type of change is appropriate in Forge before I put more work into it. (I haven't touched modding in quite a while.)

**Comment by LexManos** [comment]
The main problem is that you havent given any performance statistics or profiler data to justify these changes.
I have no idea if its possible for rendering things like this, but it'd be good if you could throw a test suite together in the test directory that actually profiles the two different methods and gives actual, reproduceable stats.

**Comment by asiekierka** [comment]
I'm still of the opinion this should be a separate mod. I do not think a
rewrite of the cloud renderer belongs in a compatibility layer/mod loader
that Forge essentially is.

2016-08-13 0:14 GMT+02:00 LexManos notifications@github.com:

**Comment by RANKSHANK** [comment]
^So encourage more community driven ASM usage on high traffic classes?

@Zaggy1024 original post says ~0.47% v ~9-11% game render time. That's quite substantial if it can be proven. Try doing what mezz did [here](https://github.com/MinecraftForge/MinecraftForge/pull/3093#issuecomment-233108423) in #3093: standardized test scenario, and outcome for standard forge and modified forge.

**Comment by LexManos** [comment]
@asiekierka your opinion doesn't matter.
This is a simple change matanance wise. Which honestly, it it goes in Forge will most likely be eaten by vanilla in a version or two.
When it comes to these types of changes, this guy is actually doing it perfectly.
Smallest possible patch to the base classes, simple configuration to enable/disable.
One comment I have on it tho, is there is no need for it to have a tick handler to count ticks.
It should be able to just pass in the world tick info from the render call like it does for partial ticks.

The only real issue is that we has no hard verification to justify the changes.

**Comment by asiekierka** [comment]
Even opinions which do not matter may be voiced - not to be considered, even, but merely to be present.If vanilla eats it up, I am for it... the devs might not notice it otherwise... But I would be wary of extra maintenance.-------- Wiadomość oryginalna --------Temat: Re: [MinecraftForge/MinecraftForge] Add a cloud renderer that uploads geometry to the GPU, rendering much faster than vanilla. (#2872)Od: LexManos Do: MinecraftForge/MinecraftForge DW: Adrian Siekierka ,Mention @asiekierka your opinion doesn't matter.
This is a simple change matanance wise. Which honestly, it it goes in Forge will most likely be eaten by vanilla in a version or two.
When it comes to these types of changes, this guy is actually doing it perfectly.
Smallest possible patch to the base classes, simple configuration to enable/disable.

The only real issue is that we has no hard verification to justify the changes.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/MinecraftForge/MinecraftForge","title":"MinecraftForge/MinecraftForge","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/MinecraftForge/MinecraftForge"}},"updates":{"snippets":[{"icon":"PERSON","message":"@LexManos in #2872: @asiekierka your opinion doesn't matter.\r\nThis is a simple change matanance wise. Which honestly, it it goes in Forge will most likely be eaten by vanilla in a version or two.\r\nWhen it comes to these types of changes, this guy is actually doing it perfectly.\r\nSmallest possible patch to the base classes, simple configuration to enable/disable.\r\n\r\nThe only real issue is that we has no hard verification to justify the changes.\r\n"}],"action":{"name":"View Pull Request","url":"https://github.com/MinecraftForge/MinecraftForge/pull/2872#issuecomment-239608011"}}}

**Comment by asiekierka** [comment]
Besides, if we're hooking the cloud renderer anyway, why not let mods provide their own cloud renderers for their own worlds? That way, the optimization itself could be an addon which doesn't use ASM, and mods would benefit as well. I'm sure many would appreciate that, even mods which don't add dimensions like CloudMaster.

**Comment by Zaggy1024** [comment]
There is actually a hook for custom cloud renderers already, if I remember correctly. It's been a long time since I touched this code, though, so I can't remember why I decided against using that to hook this renderer.

**Comment by LexManos** [comment]
Yes there is already a sky and cloud renderer, And Asi you're just sitting here pissing me off. Your opinion is not wanted. Shut up. Stop spamming my issue tracker.
I understand what you are saying, I know why you are are saying it. And I don't know how much more I can make this clear I DONT FUCKING CARE ABOUT YOUR OPINION.
Forge HAS MANY performance enhancements to the vanilla code base. We are not JUST a compatibility layer, we are a major proving platform for new technologies. 

No need for any more discussion beyond what has been requested of @Zaggy1024, if he can create a test harness that actually validates his claims of performance improvements then we can move on from there.

**Comment by asiekierka** [comment]
I'm not "sitting here pissing you off". That is not my intention at all,
rather I'm trying to voice my opinion on subjects relevant to the area of
modding. I believe that I have a right to do so, as Forge, at least in my
view, seeks to be a platform which lets all modders build under a common
base. I am also not "spamming your issue tracker" - unless comments from
modders on the various pull requests being submitted to Forge (as well as
notes to give them another look after months of inactivity) are "unwanted
messages" to you.

My worry was not that Forge is a proving platform or not, my worry was that
you generally expressed a dislike for having to maintain large amounts of
code in the past - so I pointed that out in good faith. Once you disagreed,
I gave it deeper thought, understanding that your decision to include this
in Forge is not changeable - the CloudRenderer could very well be a
replaceable hook for other modders to benefit from, not a singleton
instance, even if you keep the performance improvement merged in.

2016-08-13 21:35 GMT+02:00 LexManos <notifications@github.com>:

**Comment by Fortist** [comment]
Lex is a faggot.

**Pull request closed without merging**